src/psyclone/transformations.py

```diff
@@ -2825,3 +2824,216 @@ def validate(self, loop):
+        from psyclone.gocean1p0 import GOLoop
+
+        # Extracting distributed memory code is not supported. This
+        # constraint covers the presence of HaloExchange and GlobalSum
+        # classses as they are only generated when distributed memory
+        # is enabled.
+        if Config.get().distributed_memory:
+            raise TransformationError(
+                "Error in {0}: Distributed memory is not supported."
+                .format(str(self.name)))
+
+        # Now check constrains for individual Nodes in node list
+        for node in node_list:
+
+            # Check that an ExtractNode is not already in the node list
+            # marked for extraction. Otherwise we would have an extract
+            # region within another extract region.
+            if node.walk(node.children, ExtractNode) or \
+               isinstance(node, ExtractNode):
+                raise TransformationError(
+                    "Error in {0}: Extraction of a region which already "
+                    "contains another Extract region is not allowed."
+                    .format(str(self.name)))
+
+            # Check that ExtractNode is not inserted between a Kernel or
+            # a BuiltIn call and its parent Loop.
+            if isinstance(node, (Kern, BuiltIn)) and \
+               isinstance(node.parent, Loop):
+                raise TransformationError(
+                    "Error in {0}: Extraction of a Kernel or a Built-in "
+                    "call without its parent Loop is not allowed."
+                    .format(str(self.name)))
+
+            # Check that ExtractNode is not inserted between a Loop and its
+            # parent Directive when optimisations are applied, as this may
+            # result in including the end of Directive for extraction but
+            # not the beginning.
+            if isinstance(node, Loop) and isinstance(node.parent, Directive):
+                raise TransformationError(
+                    "Error in {0}: Extraction of a Loop without its parent "
+                    "Directive is not allowed.".format(str(self.name)))
+
+            # Check that ExtractNode is not inserted between an orphaned
+            # Directive (e.g. OMPDoDirective, ACCLoopDirective) and its
+            # ancestor Directive (e.g. ACC or OMP Parallel Directive) when
+            # optimisations are applied. Note that we need to explicitly
+            # exclude the OMPParallelDoDirective as it inherits from both
+            # OMPDoDirective and OMPParallelDirective.
+            if isinstance(node, (OMPDoDirective, ACCLoopDirective)) and \
+               not isinstance(node, OMPParallelDoDirective):
+                raise TransformationError(
+                    "Error in {0}: Extraction of an orphaned Directive "
+                    "without its ancestor Directive is not allowed."
+                    .format(str(self.name)))
+
+            # Dynamo0.3 API constraint: Check that ExtractNode is not
```

Comment:
This API-specific check should be in an API-specific sub-class.

Reply:
Created sub-classes and modified tests accordingly.

Reply:
Great.